### DISCUSSION TITLE
[04.아키텍처] 외래 키

### DISCUSSION THREAD

**Discussion by choihuk** [Learning-Is-Vital-In-Development/23-6-RealMySQL#12] [General]
(99p) 외래 키는 데이터베이스 서버 운영의 불편함 때문에 서비스용 데이터베이스에서는 생성하지 않는 경우도 있다는 말이 잘 이해가 되지 않습니다.

**Comment by tlarbals824** [comment]
https://engineering-skcc.github.io/oracle%20tuning/foreign_key_없이_구축하는_DB/

**Comment by dojinyou** [comment]
성능적인 문제
외래키 제약조건을 검사하는 과정에서 데이터 변경을 막기 위해 공유 잠금을 잡고 조회를 합니다. 또한 Cascade 조건에 의해 데이터 변경이 일어날 경우에는 배타적 잠금을 잡고 데이터를 변경을 하기 때문에 성능상의 문제가 있을 수 있습니다. 하지만 이것은 미미합니다.
ref: [공식문서](https://dev.mysql.com/doc/refman/8.0/en/create-table-foreign-keys.html)

운영상의 문제
실제로 데이터 제약 조건에 의해서 데이터를 넣는 순서가 제약되거나 즉각적인 데이터 변경이 불가능할 수 있습니다. 하지만 이것은 정합성을 보장하기 위해서 감수해야하는 적은 불편함 정도라고 생각합니다.

이러한 불편함을 해소하기 위해서 시스템 변수인 `foreign_key_checks`를 `OFF`로 만들고 작업을 처리할 수 있습니다.
